src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.cs

```diff
@@ -48,20 +52,38 @@ internal partial class WorkspaceProjectContextHost : AbstractMultiLifetimeCompon
             _activeConfiguredProjectProvider = activeConfiguredProjectProvider;
             _applyChangesToWorkspaceContextFactory = applyChangesToWorkspaceContextFactory;
             _dataProgressTrackerService = dataProgressTrackerService;
+            _languageServiceTelemetryService = languageServiceTelemetryService;
+        }
+
+        private string ProjectId
+        {
+            get
+            {
+                if (Strings.IsNullOrEmpty(_projectId))
+                {
+                    string? fullPath = _project?.UnconfiguredProject?.FullPath;
+                    _projectId = Strings.IsNullOrEmpty(fullPath) ? string.Empty : _languageServiceTelemetryService.HashValue(fullPath);
+                }
+
+                return _projectId;
+            }
```

Comment:
As I understand it, this value is only used for telemetry, but its name suggests something grander.

If the language service telemetry component was in the `ConfiguredProject` scope, it could import the `ConfiguredProject` via its constructor and cache this value itself, keeping this telemetry concept out of the language service code. It seems you could then remove `HashValue` from the interface as well.

Looking further, it seems that this usage is in the configured project, and a usage below is in the unconfigured project scope, so the approach I describe here would only work if you split that service in two. I'm less sold on the idea, but thought I'd leave this thought here in case it prompts something.

At the very least, consider renaming to something like `TelemetryId`.

Reply:
I have split the interface into 2 in [https://github.com/dotnet/project-system/commit/d718c99e5c46290d4ee63327c83188346ece0b9f](https://github.com/dotnet/project-system/commit/d718c99e5c46290d4ee63327c83188346ece0b9f) so that unconfigured and configured scopes are handled separately since the most likely change that could be needed is reporting additional telemetry and this seems like the cleaner way to handle that.

---
In reply to: [567238918](https://github.com/dotnet/project-system/pull/6880#discussion_r567238918) [](ancestors = 567238918)

Reply:
Still working on a fix for the duplication of the ProjectTelemetryId property though.

---
In reply to: [567490682](https://github.com/dotnet/project-system/pull/6880#discussion_r567490682) [](ancestors = 567490682,567238918)